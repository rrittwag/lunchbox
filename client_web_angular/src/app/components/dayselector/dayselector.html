<div class="day-control">
  <div class="row horizontal-align">
    <div class="col-xs-12 col-sm-9 col-md-6 col-lg-4">
      <div class="row vertical-align">
        <div class="col-xs-2">
          <a class="btn btn-primary glyphicon glyphicon-chevron-left"
              ng-click="goPrevDay()" ng-disabled="!hasPrevDay()"></a>
        </div>
        <div class="col-xs-8 text-center">
          <h2><small>{{selectedDay | formatToWeekday}}</small></h2>
          <h2>{{selectedDay | date:'dd.MM.yyyy'}}</h2>
        </div>
        <div class="col-xs-2">
          <a class="btn btn-primary glyphicon glyphicon-chevron-right pull-right"
              ng-click="goNextDay()" ng-disabled="!hasNextDay()"></a>
        </div>
      </div>
    </div>
  </div>
</div>
