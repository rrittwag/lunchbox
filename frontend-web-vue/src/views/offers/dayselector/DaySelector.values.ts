export enum DaySelectorDirection {
  PREVIOUS = 'prev',
  NEXT = 'next',
}

export const LABEL_GO_TO_PREVIOUS_DAY = 'Zu vorherigem Tag wechseln'
export const LABEL_GO_TO_NEXT_DAY = 'Zu n√§chstem Tag wechseln'
