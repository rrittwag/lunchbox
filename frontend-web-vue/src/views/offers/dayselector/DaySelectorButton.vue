<script setup lang="ts">
import {
  DaySelectorDirection,
  LABEL_GO_TO_NEXT_DAY,
  LABEL_GO_TO_PREVIOUS_DAY,
} from '@/views/offers/dayselector/DaySelector.values'
import AngleLeftIcon from '~icons/fa-solid/angle-left'
import AngleRightIcon from '~icons/fa-solid/angle-right'

const props = defineProps<{
  disabled?: boolean
  direction: DaySelectorDirection
}>()
const emit = defineEmits<{
  (e: 'click'): void
}>()

const onClick = () => emit('click')

const isPrevious = props.direction === DaySelectorDirection.PREVIOUS
const title = isPrevious ? LABEL_GO_TO_PREVIOUS_DAY : LABEL_GO_TO_NEXT_DAY
</script>

<template>
  <button
    class="cursor-pointer origin-center text-primary-500 hover:scale-110 hover:text-primary-600 disabled:cursor-not-allowed disabled:text-primary-400 disabled:opacity-25"
    :class="{ 'active:text-primary-800': !props.disabled }"
    :title="title"
    :aria-label="title"
    :disabled="props.disabled"
    :aria-keyshortcuts="isPrevious ? 'ArrowLeft' : 'ArrowRight'"
    @click="onClick"
  >
    <AngleLeftIcon v-if="isPrevious" class="h-16 w-16" aria-hidden="true" />
    <AngleRightIcon v-else class="h-16 w-16" aria-hidden="true" />
  </button>
</template>
