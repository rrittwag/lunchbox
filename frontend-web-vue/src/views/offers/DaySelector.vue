<template>
  <div class="flex justify-between items-center">
    <h2 class="grow text-center block text-xl text-neutral-800">
      {{ selectedDayAsWeekday }}
      <time class="block text-lg font-light text-neutral-700" :datetime="selectedDayAsISOString">
        {{ selectedDayAsDateString }}
      </time>
    </h2>

    <DaySelectorButton
      class="order-first"
      :direction="DaySelectorDirection.PREVIOUS"
      :disabled="props.disabledPrev"
      @click="onClickPrev"
    />
    <DaySelectorButton
      :direction="DaySelectorDirection.NEXT"
      :disabled="props.disabledNext"
      @click="onClickNext"
    />
  </div>
</template>

<script setup lang="ts">
import { computed, defineEmits, defineProps, onMounted, onUnmounted } from 'vue'
import { formatToISODate, formatToLocalDate, formatToWeekday } from '@/util/formatting'
import DaySelectorButton from '@/views/offers/dayselector/DaySelectorButton.vue'
import { DaySelectorDirection } from '@/views/offers/dayselector/DaySelectorDirection'

const props = defineProps<{
  selectedDay: Date
  disabledPrev?: boolean
  disabledNext?: boolean
}>()
const emit = defineEmits<{
  (e: 'change', direction: DaySelectorDirection): void
}>()

const selectedDayAsWeekday = computed(() => formatToWeekday(props.selectedDay))
const selectedDayAsDateString = computed(() => formatToLocalDate(props.selectedDay))
const selectedDayAsISOString = computed(() => formatToISODate(props.selectedDay))

const onClickPrev = () => emit('change', DaySelectorDirection.PREVIOUS)
const onClickNext = () => emit('change', DaySelectorDirection.NEXT)

onMounted(() => window.addEventListener('keydown', onKeydown))
onUnmounted(() => window.removeEventListener('keydown', onKeydown))

function onKeydown(event: KeyboardEvent) {
  if (event.defaultPrevented) {
    return // Do nothing if the event was already processed
  }

  if (event.altKey || event.ctrlKey || event.shiftKey || event.metaKey) return

  switch (event.key) {
    case 'Left': // IE/Edge specific value
    case 'ArrowLeft':
      onClickPrev()
      break
    case 'Right': // IE/Edge specific value
    case 'ArrowRight':
      onClickNext()
      break
    default:
      return // Quit when this doesn't handle the key event.
  }

  // Cancel the default action to avoid it being handled twice
  event.preventDefault()
}
</script>
