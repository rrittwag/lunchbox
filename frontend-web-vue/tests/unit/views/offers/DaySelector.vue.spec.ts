/* eslint-disable @typescript-eslint/no-non-null-assertion */
import DaySelector from '@/views/offers/DaySelector.vue'
import { DaySelectorDirection } from '@/views/offers/dayselector/DaySelectorDirection'
import { mount } from '@vue/test-utils'

describe('DaySelector', () => {
  it('renders snapshot', () => {
    const wrapper = mount(DaySelector, {
      props: { selectedDay: YESTERDAY },
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('has disabled previous button', () => {
    const wrapper = mount(DaySelector, {
      props: { selectedDay: YESTERDAY, disabledPrev: true },
    })

    const buttons = wrapper.findAll('button')
    expect(buttons[0].attributes()).toHaveProperty('disabled')
    expect(buttons[1].attributes()).not.toHaveProperty('disabled')
  })

  it('has disabled next button', () => {
    const wrapper = mount(DaySelector, {
      props: { selectedDay: TOMORROW, disabledNext: true },
    })

    const buttons = wrapper.findAll('button')
    expect(buttons[0].attributes()).not.toHaveProperty('disabled')
    expect(buttons[1].attributes()).toHaveProperty('disabled')
  })

  it('emits event WHEN click previous day', async () => {
    const wrapper = mount(DaySelector, {
      props: { selectedDay: TODAY },
    })

    await wrapper.findAll('button')[0].trigger('click')

    expect(wrapper.emitted().change).toBeDefined()
    expect(wrapper.emitted().change!.length).toBe(1)
    expect(wrapper.emitted().change![0]).toEqual([DaySelectorDirection.PREVIOUS])
  })

  it('emits event WHEN click next day', () => {
    const wrapper = mount(DaySelector, {
      props: { selectedDay: TODAY },
    })

    wrapper.findAll('button')[1].trigger('click')

    expect(wrapper.emitted().change).toBeDefined()
    expect(wrapper.emitted().change!.length).toBe(1)
    expect(wrapper.emitted().change![0]).toEqual([DaySelectorDirection.NEXT])
  })

  it('emits event WHEN pressing left arrow key', () => {
    const wrapper = mount(DaySelector, {
      props: { days: [YESTERDAY, TODAY, TOMORROW], selectedDay: TODAY },
    })

    window.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowLeft' }))

    expect(wrapper.emitted().change).toBeDefined()
    expect(wrapper.emitted().change!.length).toBe(1)
    expect(wrapper.emitted().change![0]).toEqual([DaySelectorDirection.PREVIOUS])
  })

  it('emits event WHEN pressing right arrow key', () => {
    const wrapper = mount(DaySelector, {
      props: { days: [YESTERDAY, TODAY, TOMORROW], selectedDay: TODAY },
    })

    window.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowRight' }))

    expect(wrapper.emitted().change).toBeDefined()
    expect(wrapper.emitted().change!.length).toBe(1)
    expect(wrapper.emitted().change![0]).toEqual([DaySelectorDirection.NEXT])
  })
})

// mocks 'n' stuff

const YESTERDAY = new Date('2019-12-01')
const TODAY = new Date('2019-12-02')
const TOMORROW = new Date('2019-12-03')
