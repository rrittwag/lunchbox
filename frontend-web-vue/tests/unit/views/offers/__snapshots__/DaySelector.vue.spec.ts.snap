// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DaySelector renders snapshot 1`] = `
<div
  class="flex justify-between items-center"
>
  <h2
    class="grow text-center block text-xl text-neutral-800"
  >
    Sonntag 
    <time
      class="block text-lg font-light text-neutral-700"
      datetime="2019-12-01"
    >
      1.12.2019
    </time>
  </h2>
  <button
    aria-keyshortcuts="ArrowLeft"
    aria-label="Zu vorherigem Tag wechseln"
    class="text-primary-500 hover:text-primary-600 origin-center hover:scale-110 disabled:text-primary-400 disabled:opacity-25 disabled:cursor-not-allowed active:text-primary-800 order-first"
    title="Zu vorherigem Tag wechseln"
  >
    <!---->
  </button>
  <button
    aria-keyshortcuts="ArrowRight"
    aria-label="Zu nächstem Tag wechseln"
    class="text-primary-500 hover:text-primary-600 origin-center hover:scale-110 disabled:text-primary-400 disabled:opacity-25 disabled:cursor-not-allowed active:text-primary-800"
    title="Zu nächstem Tag wechseln"
  >
    <!---->
  </button>
</div>
`;
